### PR TITLE
feat: Undo/Redo

### LINKED ISSUE #3915: [Enhancement] Platform change to include the DSL in a data structure which allows for tracking and merging changes
We should be handling the undo-redo states from a separate thread to avoid blocking the UI thread while the undo-able DSL is being updated.

Libraries which can be used:
- [yjs](https://github.com/yjs/yjs)
- (optional) Redux in workers to store the DSL
- (optional) Store state in [indexeddb](https://github.com/yjs/y-indexeddb)

Derisk
====
- [ ] yjs undo manager supports redo
- [ ] Performance of undo redo updates on DSL
- [ ] Performance optimizations by using diffs for DSL updates
- [ ] Setting up yjs-indexeddb for persisting state. This includes cache invalidating from the servers.

Tasks
====
- [ ] Setup yjs data structure on page load
- [ ] Function to convert DSL update paths to yjs data structure updates
- [ ] Function to convert yjs undone changes to DSL update paths

### LINKED ISSUE #3916: [Enhancement]  Undo and Redo operation functions
- [ ] Function which undo's the structure which maintains undo-able state.
- [ ] Function which redo's the structure which maintains redo-able state.
- [ ] Sagas to trigger undo/redo
- [ ] Action payload definitions to trigger undo/redo
- [ ] Global hot keys for undo/redo

### LINKED ISSUE #3918: [Enhancement] Unit tests for undo redo methods

### LINKED ISSUE #3919: [Enhancement] Cypress tests for Undo Redo

### LINKED ISSUE #5058: Composite operations
Some operations today send multiple independent updates to the DSL. We need to make sure that these operations now become atomic, but using the `batchUpdateWidgetProperty` feature.

Operations of note:
- [ ] Move widget - Updates main container height and widget positions
- [ ] Add widget - Updates main container height and widget positions
- [ ] Make widget resize operation atomic
- [ ] Make property updates from property pane atomic
- [ ] Add arch to allow multi widget / multi property updates from property control
- [ ] Make Tabs operations(Add/Remove/Tab name change) atomic.

### PR DESCRIPTION
**Pull request by rahulramesha**
## Description

Undo/redo feature for Editor

Fixes #3915 
Fixes #3916  
Fixes #3918 
Fixes #3919 

Atomic Operations - Making DSL operations atomic(single DSL update on wire as well as on the app state)

- [x] De-coupled a few widget operations from WidgetOperationSagas(which was becoming too big)
- [x] Make widget resize operation atomic
- [x] Make property updates from property pane atomic
- [x] Add arch to allow multi widget / multi property updates from property control
- [x] Make Tabs operations(Add/Remove/Tab name change) atomic.

Fixes #5058

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
Manual Test, Jest Tests and Cypress Tests

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: undo_redo_poc 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.61 **(-0.09)** | 36.39 **(-0.09)** | 33.83 **(-0.02)** | 55.2 **(-0.04)**
 :red_circle: | app/client/src/actions/controlActions.tsx | 64.71 **(-12.21)** | 0 **(-100)** | 0 **(-25)** | 50 **(-16.67)**
 :green_circle: | app/client/src/actions/pageActions.tsx | 70.59 **(1.2)** | 25 **(0)** | 33.33 **(4.76)** | 61.43 **(2.34)**
 :red_circle: | app/client/src/components/ads/ColorPickerComponent.tsx | 35 **(-2.84)** | 53.85 **(-9.79)** | 0 **(0)** | 36.84 **(-3.16)**
 :green_circle: | app/client/src/components/ads/DraggableList.tsx | 44.26 **(3.28)** | 33.33 **(0)** | 21.43 **(0)** | 51.02 **(4.08)**
 :red_circle: | app/client/src/components/ads/DraggableListComponent.tsx | 43.75 **(-18.32)** | 50 **(-50)** | 37.5 **(4.17)** | 46.15 **(-23.42)**
 :red_circle: | app/client/src/components/ads/Toast.tsx | 77.55 **(-1.17)** | 63.33 **(0.83)** | 83.33 **(-4.91)** | 77.55 **(-1.17)**
 :red_circle: | app/client/src/components/editorComponents/DropTargetComponent.tsx | 88.73 **(-1.01)** | 68.63 **(0.71)** | 90 **(-0.91)** | 88.41 **(-1.06)**
 :red_circle: | app/client/src/components/editorComponents/DropTargetUtils.ts | 90 **(-10)** | 66.67 **(0)** | 100 **(0)** | 88.89 **(-11.11)**
 :red_circle: | app/client/src/components/editorComponents/EditorContextProvider.tsx | 92.86 **(-0.47)** | 100 **(0)** | 66.67 **(0)** | 92.86 **(-0.47)**
 :green_circle: | app/client/src/components/editorComponents/ResizableComponent.tsx | 57.94 **(-0.24)** | 20.93 **(0.48)** | 66.67 **(0)** | 58.49 **(0.55)**
 :red_circle: | app/client/src/components/editorComponents/CodeEditor/index.tsx | 67.41 **(-0.5)** | 39.75 **(-1.01)** | 52.5 **(0)** | 68.15 **(-0.55)**
 :green_circle: | app/client/src/components/propertyControls/InputTextControl.tsx | 78.95 **(1.17)** | 28.57 **(0)** | 66.67 **(0)** | 77.78 **(1.31)**
 :red_circle: | app/client/src/components/propertyControls/KeyValueComponent.tsx | 17.14 **(-0.6)** | 29.63 **(0)** | 0 **(0)** | 18.75 **(-0.22)**
 :red_circle: | app/client/src/components/propertyControls/MenuItemsControl.tsx | 22.83 **(-4.8)** | 29.41 **(-10.59)** | 7.69 **(-1.4)** | 24.69 **(-3.09)**
 :red_circle: | app/client/src/components/propertyControls/PrimaryColumnsControl.tsx | 20.35 **(-1.98)** | 12.5 **(-2.05)** | 7.14 **(-0.86)** | 21.57 **(-1.35)**
 :green_circle: | app/client/src/components/propertyControls/SwitchControl.tsx | 92.86 **(0)** | 80 **(13.33)** | 75 **(0)** | 91.67 **(0)**
 :red_circle: | app/client/src/components/propertyControls/TabControl.tsx | 40.4 **(-0.3)** | 40 **(-8)** | 22.22 **(1.39)** | 42.7 **(2.46)**
 :red_circle: | app/client/src/constants/messages.ts | 74.62 **(-0.04)** | 100 **(0)** | 24.31 **(-0.12)** | 80.87 **(-0.23)**
 :green_circle: | app/client/src/pages/Editor/GlobalHotKeys.tsx | 65.14 **(0.99)** | 30.3 **(0)** | 56.41 **(2.36)** | 63.81 **(1.06)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Widgets/WidgetContextMenu.tsx | 76.32 **(0.71)** | 16.67 **(0)** | 66.67 **(0)** | 75.68 **(0.68)**
 :red_circle: | app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx | 50.75 **(-6.64)** | 32.95 **(-7.33)** | 21.43 **(-3.57)** | 51.59 **(-7.29)**
 :green_circle: | app/client/src/pages/Editor/PropertyPane/PropertySection.tsx | 86.21 **(0.5)** | 57.14 **(0)** | 66.67 **(0)** | 85.19 **(0.57)**
 :green_circle: | app/client/src/sagas/DraggingCanvasSagas.ts | 84.29 **(8.68)** | 39.29 **(15.76)** | 100 **(0)** | 84.62 **(7.7)**
 :green_circle: | app/client/src/sagas/EvaluationsSaga.ts | 24.38 **(3.67)** | 6.25 **(4.69)** | 13.64 **(3.64)** | 27.65 **(3.84)**
 :sparkles: :new: | **app/client/src/sagas/ReplaySaga.ts** | **31.11** | **4.17** | **33.33** | **35**
 :red_circle: | app/client/src/sagas/SnipingModeSagas.ts | 21.21 **(-0.79)** | 3.33 **(0)** | 40 **(0)** | 22.11 **(-0.81)**
 :sparkles: :new: | **app/client/src/sagas/WidgetAdditionSagas.ts** | **69.34** | **60** | **62.5** | **70.25**
 :sparkles: :new: | **app/client/src/sagas/WidgetDeletionSagas.ts** | **38.51** | **20.24** | **38.1** | **39.86**
 :red_circle: | app/client/src/sagas/WidgetOperationSagas.tsx | 42.35 **(-9.98)** | 35.75 **(-1.63)** | 39.58 **(-12.37)** | 44.26 **(-8.76)**
 :green_circle: | app/client/src/sagas/index.tsx | 92.59 **(0.14)** | 60 **(0)** | 100 **(0)** | 94.23 **(0.11)**
 :green_circle: | app/client/src/utils/DSLMigrations.ts | 76.61 **(1.57)** | 66.8 **(0.6)** | 69.77 **(0)** | 76.26 **(1.62)**
 :red_circle: | app/client/src/utils/DynamicBindingUtils.ts | 87.76 **(-1.9)** | 77.22 **(-6.32)** | 73.91 **(-4.35)** | 86.05 **(-2.14)**
 :green_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 86.9 **(0)** | 76 **(2)** | 84.62 **(0)** | 83.58 **(0)**
 :red_circle: | app/client/src/utils/helpers.tsx | 64.71 **(-0.44)** | 36.89 **(-2.29)** | 45.45 **(-1.43)** | 62.11 **(-0.07)**
 :sparkles: :new: | **app/client/src/utils/replayHelpers.tsx** | **45.65** | **23.08** | **14.29** | **45.95**
 :red_circle: | app/client/src/utils/storage.ts | 33.75 **(-1.25)** | 16.67 **(3.03)** | 9.09 **(-3.73)** | 26.05 **(-1.77)**
 :red_circle: | app/client/src/utils/hooks/useBlocksToBeDraggedOnCanvas.ts | 95.41 **(-0.32)** | 84.38 **(-1.33)** | 85.71 **(-0.65)** | 95.24 **(-0.26)**
 :red_circle: | app/client/src/widgets/BaseWidget.tsx | 73.98 **(-0.61)** | 37.78 **(-1.75)** | 68.75 **(0)** | 72.88 **(-0.62)**
 :red_circle: | app/client/src/widgets/TabsWidget/index.ts | 31.25 **(-11.61)** | 0 **(-100)** | 0 **(0)** | 31.25 **(-11.61)**
 :green_circle: | app/client/src/widgets/TabsWidget/widget/index.tsx | 77.17 **(10.02)** | 48.53 **(2.99)** | 78.95 **(18.95)** | 77.01 **(10.34)**
 :sparkles: :new: | **app/client/src/workers/ReplayDSL.ts** | **91.8** | **76.47** | **87.5** | **91.38**
 :sparkles: :new: | **app/client/src/workers/replayUtils.ts** | **91.3** | **75** | **100** | **95.12**</details>

### CODE REVIEW

**Review by jsartisan — DISMISSED** [review]

File: app/client/src/components/ads/Toast.tsx [hunk 1]
```diff
@@ -115,6 +116,10 @@ const StyledDebugButton = styled(DebugButton)`
   margin-left: auto;
 `;
 
+const StyledActionText = styled(Text)`
+  color: #f8682b !important;
```
Comment (rahulramesha) [inline app/client/src/components/ads/Toast.tsx:120]: move to theme file

File: app/client/src/sagas/ReplaySaga.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+export default function* undoRedoListenerSaga() {
+  yield takeEvery(ReduxActionTypes.UNDO_REDO_OPERATION, undoRedoSaga);
+}
+
+export function* openPropertyPaneSaga(replay: any) {
+  const replayWidgetId = Object.keys(replay.widgets)[0];
+
+  if (!replayWidgetId || !replay.widgets[replayWidgetId].propertyUpdates)
```
Comment (rahulramesha) [inline app/client/src/sagas/ReplaySaga.ts:36]: Add comments

File: app/client/src/sagas/ReplaySaga.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    return;
+
+  scrollWidgetIntoView(replayWidgetId);
+
+  const isPropertyPaneVisible = yield select(getIsPropertyPaneVisible);
```
Comment (rahulramesha) [inline app/client/src/sagas/ReplaySaga.ts:41]: add types

File: app/client/src/sagas/WidgetOperationSagas.tsx [hunk 4]
```diff
@@ -375,6 +367,22 @@ export function* addChildSaga(addChildAction: ReduxAction<WidgetAddChild>) {
   }
 }
 
+// Page load -> DSL -> unevalTree -> yjs structure
```
Comment (rahulramesha) [inline app/client/src/sagas/WidgetOperationSagas.tsx:370]: Remove this

File: app/client/src/utils/replayHelpers.tsx [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+export const showUndoRedoToast = (
+  widgetName: string | undefined,
+  isMultiple: boolean,
+  isCreated: boolean,
+  shouldUndo: boolean,
+) => {
+  if (shouldDisallowToast(shouldUndo)) return;
+
+  const actionDescription = isCreated
```
Comment (rahulramesha) [inline app/client/src/utils/replayHelpers.tsx:65]: move to diff function and use switch case

File: app/client/src/utils/replayHelpers.tsx [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+/**
+ * search the dom with id of element and scroll the page to its position
+ *
+ * @param id
+ */
+export const scrollWidgetIntoView = (id: string) => {
+  const el = document.getElementById(id);
+  if (el)
+    scrollIntoView(el, {
```
Comment (rahulramesha) [inline app/client/src/utils/replayHelpers.tsx:90]: use element's methopd instead

File: app/client/src/workers/ReplayDSL.ts [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+import { captureException } from "@sentry/react";
+import { diff as deepDiff, applyChange, revertChange } from "deep-diff";
+
+import { processDiff, DSLDiff, getPathsFromDiff } from "./replayUtils";
+import { CanvasWidgetsReduxState } from "reducers/entityReducers/canvasWidgetsReducer";
+
+const _DIFF_ = "diff";
+
+export default class ReplayDSL {
+  diffMap: any;
```
Comment (rahulramesha) [inline app/client/src/workers/ReplayDSL.ts:11]: convert to private whereever reqd

File: app/client/src/workers/ReplayDSL.ts [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+  undo() {
+    const start = performance.now();
+    const diffs = this.getDiffs();
+
+    if (this.shouldReplay(true)) {
+      this.undoManager.undo();
+      const replay = this.applyDiffs(diffs, true);
+      const stop = performance.now();
+
+      this.logs.push({
```
Comment (rahulramesha) [inline app/client/src/workers/ReplayDSL.ts:61]: convert to a perf log

File: app/client/src/workers/ReplayDSL.ts [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+  /**
+   * apply the diff on the current dsl
+   *
+   * @param diffs
+   * @param diffUpdate
+   */
+  applyDiffs(diffs: Array<DSLDiff>, isUndo: boolean) {
+    const replay = {};
+    const diffUpdate = isUndo ? revertChange : applyChange;
```
Comment (rahulramesha) [inline app/client/src/workers/ReplayDSL.ts:147]: change name

File: app/client/src/workers/dslReplay.worker.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
+const workerContext: Worker = self as any;
+
+export enum ReplayOperation {
```
Comment (rahulramesha) [inline app/client/src/workers/dslReplay.worker.ts:3]: remove file

File: app/client/src/workers/replayUtils.ts [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  dsl: CanvasWidgetsReduxState,
+  diff: DSLDiff,
+  replay: any,
+  isUndo: boolean,
+) {
+  if (!diff || !diff.path || !diff.path.length || diff.path[0] === "0") return;
+
+  const widgetId = diff.path[0];
+
+  switch (diff.kind) {
```
Comment (rahulramesha) [inline app/client/src/workers/replayUtils.ts:39]: add comments to understand better

**Review by jsartisan — APPROVED** [review]

File: app/client/src/utils/replayHelpers.tsx (hunk 6 above)
Comment (rahulramesha) [reply app/client/src/utils/replayHelpers.tsx:103]: turns out element's `scrollIntoViewIfNeeded` is not supported in firefox so we'll still be using third party library

**Review by jsartisan — DISMISSED** [review]

File: app/client/src/components/ads/TextInput.tsx [hunk 12]
```diff
@@ -48,6 +49,8 @@ export type TextInputProps = CommonComponentProps & {
   dataType?: string;
   theme?: any;
   rightSideComponent?: React.ReactNode;
+  onBlur?: typeof noop;
```
Comment (riodeuno) [inline app/client/src/components/ads/TextInput.tsx:52]: Are these necessary changes for this project?

**Review by marks0351 — APPROVED** [review]
approving to run cypress tests.

**Review by marks0351 — CHANGES_REQUESTED** [review]

**Review by marks0351 — APPROVED** [review]

**Review by marks0351 — APPROVED** [review]

**Review by marks0351 — DISMISSED** [review]
approving for tests

**Review by marks0351 — APPROVED** [review]

**Review by marks0351 — APPROVED** [review]

**Review by jsartisan — DISMISSED** [review]

**Comment by YogeshJayaseelan** [comment]
@rahulramesha We have identified the below issues and still testing in-progress

1.Adding two widgets to canvas and renaming one of the widget and keeping the property pane open and press cmd+z to undo and cmd+shift+z to redo then close the property pane and press cmd+z - The widgets get selected and ready to be grouped

https://www.loom.com/share/90ae610db24b4a2d8fca55c4ed20400b

2. Choosing any of the options in checkbox group widget individually and clicking on cmd+z doesn't undo the last selection in the checkbox group widget. whereas if the user drags few widgets along with checkbox group widget to canvas and then grouping all the widget to an container and making selections in the checkbox group and clicking on cmd+d undo's the checkbox group selection

https://www.loom.com/share/4e2f9de2a1784b20bcabce291dc285d3

**Comment by rahulramesha** [comment]
@YogeshJayaseelan pushed a fix for the first issue, 
for second issue, we don't have support for undo-redo for widget meta properties as it is meant for end users

**Comment by rahulramesha** [comment]
@YogeshJayaseelan
issue 3, Whenever Input field is in focus, undo-redo is handled by that particular input field itself, this is as expected.
issue 4, Undo-redo does not work on widget meta properties. I agree that the UX in this corner case is not great but we don't have a readily available solution for this.
issues 5 and 6 were fixed.

**Comment by YogeshJayaseelan** [comment]
@rahulramesha Verified the fixes and please find the status below,

Issues fixed:
1,5

Agreed on :
2 - Meta property is not included
3 - Undo/Redo will focus on input field change 
4- Meta property is not included

Reopen:
6 - Still exists - Changing the modal type to form and using cmd+z for the first time the modal is not defaulted to alert modal 

https://www.loom.com/share/4cc6171018834fb29df1285dd7ad99a9

Suggestion:
7 - Can we implement this feature for API and Datasource

**Comment by rahulramesha** [comment]
@YogeshJayaseelan please test 6 again, with the latest resizable modal gone in, this wont be a problem anymore.

**Comment by YogeshJayaseelan** [comment]
@rahulramesha Verified the issue 6 and now there is no options to change modal type and problem is resolved. We can proceed further

**Review by jsartisan — DISMISSED** [review]

**Review by marks0351 — APPROVED** [review]

**Review by jsartisan — APPROVED** [review]

**Pull request merged**